docker.js

```diff
@@ -5,92 +5,212 @@ const fse = require('fs-extra');
-    const imgStream = await docker.image.build(tarStream, { t: imageName });
-    // Set up logging, and catch errors, for the image build stream:
-    await promisifyStream(imgStream, 'image', snippetId);
+  try {
+    /*
+    (A) Set up temp directory of files for docker,
+    and a way to clean up this temp dir.
+
+    Based on the indexContents passed in on index.js we have a tmpDir string,
+    and a tmpDirCreate function which returns a promise which resolves to the
+    following:
+      tmpDir: `/tmp/${randomName}` --> path to the directory used to build the docker image
+      tmpDirCreate: () => Promise.all(promises) --> Function that returns an array of promises. Each promise in the array is an invocation of fse.outputFile. fse.outputFile takes a file and text we want to write to that file. When fse.outputFile resolves, this means that the text we passed in has been written to the filename we passed in.
+    */
+    const tmpDirInit = makeTempDir({ dockerSnippetId, indexContents });
+    tmpDir = tmpDirInit.tmpDir;
+
+    cleanUpFunc = () => rimraf(tmpDir, () =>
+      /*
+      The callback here runs the equivalent to `rm -rf tmpDir/`;
+      i.e., recursively "force remove" all files from this directory.
+      */
+      console.warn(`Deleted temporary directory ${tmpDir}.`)
+    );
+
+    // Wait for the temporary directory to be created:
+    await tmpDirInit.tmpDirCreate();
+  } catch (err) {
+    console.error(`Error creating temporary directory: ${err}`);
+    return null;
+  }
 
 
-    // (3) go from image to container, run that container
+  try {
+    /*
+    (B) Pack the temporary directory into a zip-like "tar" stream.
+    Docker best practice is to "archive" the directory like so.
+    "Archive a directory" = pack everything into a single file,
+    so we can unpack later. Analogy: a casette tape being wound up.
+    */
+    tarStream = await tar.pack(tmpDir);
+  } catch (err) {
+    console.error(`Error creating streamable tar file: ${err}`);
+    return null;
+  }
 
-    // Create a container (running instance) based on the image
 
-    container = await docker.container.create({ Image: imageName, name: imageName });
-    // console.log('(1) telling docker to create a container');
-    // Start that container - this is how user code is run:
+  try {
+    /*
+    (C) Build and stream an image, and create a container based
+    upon this image.
+    */
+    container = await buildImageAndCreateContainer(
+      tarStream, imageName, snippetId, docker
+    );
+
+    /*
+    (D) Start the container (in order to run user code), and
+    stream its logs.
+    */
     container = await container.start();
-    // console.log('(2) telling docker container to execute code');
-    // Get a stream (promise) for the container's logs
-    const containerStream = await container.logs({ follow: true, stdout: true, stderr: true });
+    const containerStream = await container.logs({
+      follow: true, stdout: true, stderr: true
```

Comment:
Hmm, I feel like this might be better on multiple lines? I'm not sure. AirBnB doesn't say anything about it.